apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,726 @@
+    view: 'ConversationView',
+    previous: 'thread-list',
+    partOf: 'thread'
+  }),
+  Object.freeze({
+    name: 'group-view',
+    view: 'GroupView',
+    previous: 'thread',
+    partOf: 'thread'
+  })
+]);
+
+/**
+  * OA_VIEWS holds the configuration for the old non-split architecture.
+  */
+const OA_VIEWS = Object.freeze([
+  Object.freeze({
+    name: 'thread-list',
+    url: 'index.html',
+    view: 'InboxView',
+    alsoContains: ['thread', 'composer', 'group-view', 'report-view']
+  }),
+  Object.freeze({
+    name: 'thread',
+    view: 'ConversationView',
+    previous: 'thread-list',
+    partOf: 'thread-list'
+  }),
+  Object.freeze({
+    name: 'composer',
+    view: 'ConversationView',
+    previous: 'thread-list',
+    partOf: 'thread-list'
+  }),
+  Object.freeze({
+    name: 'group-view',
+    view: 'GroupView',
+    previous: 'thread',
+    partOf: 'thread-list'
+  }),
+  Object.freeze({
+    name: 'report-view',
+    view: 'ReportView',
+    previous: 'thread',
+    partOf: 'thread-list'
+  })
+]);
+
+const VIEWS = isUsingOldArchitecture() ? OA_VIEWS : NGA_VIEWS;
+
+var currentState;
+var navigationTransition;
+
+/**
+ * Navigate to this URL.
+ *
+ * @param {String} url URL to load.
+ */
+function setLocation(url) {
+  window.location.assign(url);
+}
+
+/**
+ * We're using the old architecture if we're not in a split document.
+ *
+ * @returns {Boolean} true if we're using the old architecture.
+ */
+function isUsingOldArchitecture() {
+  var pathname = window.location.pathname;
+  return (pathname === '/' || pathname === '/index.html');
+}
+
+/**
+ * Tries to find a view name from a URL hash.
+ * It takes the part of the hash between '/' or the start of the string, and '?'
+ * or the end of the string. If the resulting string is the name of a view, it
+ * then returns this view.
+ *
+ * @param {String} hash The string to analyze.
+ * @returns {Object} The found panel, or null.
+ */
+function findViewFromHash(hash) {
+  if (!hash) {
+    return null;
+  }
+
+  var index = hash.indexOf('/');
+  if (index != -1) {
+    hash = hash.slice(index + 1);
+  }
+
+  index = hash.indexOf('?');
+  if (index !== -1) {
+    hash = hash.slice(0, index);
+  }
+
+  return findViewFromName(hash);
+}
+
+/**
+ * Tries to find the correct view from the location.
+ * First it tries to find the view using the location's hash; if
+ * findViewFromHash returns something, then we check whether the found view is
+ * indeed part of this document usinf the full URL.
+ * If findViewFromHash returns nothing, we try to find a view from the full URL.
+ *
+ * @param {Location} location The content of window.location.
+ * @returns {Object} The found view, or null.
+ */
+function findViewFromLocation(location) {
+  var pathName = location.pathname;
+  if (pathName.endsWith('/')) {
+    pathName += 'index.html';
+  }
+
+  var viewFromHash = findViewFromHash(location.hash);
+  var viewFromPath = VIEWS.find(
+    (object) => pathName.endsWith('/' + object.url)
```

Comment:
> nit: object ----> view ?

for me, it's a "view object" :) that's why I used "object". But I can use "view" just as well